paddle/cinn/optim/trans_buffer_with_dynamic_shape.cc

```diff
@@ -14,73 +14,86 @@
 
 #include "paddle/cinn/optim/trans_buffer_with_dynamic_shape.h"
 
+#include <numeric>
+#include <unordered_set>
+
 #include "paddle/cinn/common/cas.h"
+#include "paddle/cinn/common/dev_info_manager.h"
+#include "paddle/cinn/common/integer_set.h"
 #include "paddle/cinn/common/ir_util.h"
 #include "paddle/cinn/ir/ir_mutator.h"
+#include "paddle/cinn/ir/op/ir_operators.h"
+#include "paddle/cinn/ir/utils/ir_compare.h"
 #include "paddle/cinn/ir/utils/ir_copy.h"
 #include "paddle/cinn/utils/string.h"
 
 namespace cinn::optim {
 
 namespace {
 
-struct RemoveSymbolMutator : public ir::IRMutator<> {
-  explicit RemoveSymbolMutator(bool use_upper_bound) {
-    this->use_upper_bound = use_upper_bound;
-  }
-
-  void operator()(Expr* x) { ir::IRMutator<>::Visit(x, x); }
-
-  using ir::IRMutator<>::Visit;
-
-  void Visit(const ir::_Var_* var, Expr* expr) override {
-    auto node = expr->As<ir::_Var_>();
-    CHECK(node->lower_bound.defined() && node->lower_bound.is_constant() &&
-          node->upper_bound.defined() && node->upper_bound.is_constant())
-        << "Temporary buffer with dynamic shape cannot be transformed into "
-           "buffer with static shape!\n";
-    if (use_upper_bound) {
-      *expr = Expr(node->upper_bound.as_int32());
-    } else {
-      *expr = Expr(node->lower_bound.as_int32());
-    }
-  }
-
-  bool use_upper_bound;
-};
-
-RemoveSymbolMutator max_mutator(true);
-RemoveSymbolMutator min_mutator(false);
+common::cas_intervals_t var_intervals = {};
+cinn::common::SymbolicExprAnalyzer analyzer(var_intervals);
 
 struct Mutator : public ir::IRMutator<> {
   using ir::IRMutator<>::Visit;
 
+  Mutator() : shared_mem_size_used_(0) {}
+
   void Visit(const ir::_Tensor_* tensor, Expr* expr) override {
-    bool need_trans = false;
-    if (cinn::utils::Endswith(tensor->name, "_temp_buffer")) need_trans = true;
-    for (auto& e : expr->as_tensor()->shape) {
-      if (!e.is_constant()) {
-        auto new_shape1 = ir::ir_utils::IRCopy(e);
-        auto new_shape2 = ir::ir_utils::IRCopy(e);
-        max_mutator(&new_shape1);
-        min_mutator(&new_shape2);
-        new_shape1 = common::AutoSimplify(new_shape1);
-        new_shape2 = common::AutoSimplify(new_shape2);
-        CHECK(new_shape1.is_constant());
-        CHECK(new_shape2.is_constant());
-        int res_shape = std::max(new_shape1.as_int32(), new_shape2.as_int32());
-        e = ir::Expr(res_shape);
+    if (!tensor->buffer.defined()) return;
+    auto buf = tensor->buffer.As<ir::_Buffer_>();
+    if (!visited_buf_.count(buf->name)) {
+      visited_buf_.insert(buf->name);
+      auto buf_size = ir::Expr(1);
+
+      for (size_t i = 0; i < expr->as_tensor()->shape.size(); ++i) {
+        auto e = expr->as_tensor()->shape[i];
+        auto buf_e = buf->shape[i];
+        CHECK(ir::ir_utils::IRCompare(e, buf_e));
```

Comment:
Is this necessary correct? I am not sure.

Reply:
Sometimes shape of tensor is not the same as it's buffer because of reshape operation, maybe I should include situation like this.